File: app/src/main/java/com/gigaworks/tech/bible/adapter/BookAdapter.java
package com.gigaworks.tech.bible.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.gigaworks.tech.bible.R;
import com.gigaworks.tech.bible.container.Book;

import java.util.ArrayList;

public class BookAdapter extends RecyclerView.Adapter<BookAdapter.BookViewHolder> {

    private ArrayList<Book> bookList;
    private OnBookClickListener listener;

    public BookAdapter(ArrayList<Book> bookList, OnBookClickListener listener) {
        this.bookList = bookList;
        this.listener = listener;
    }

    public interface OnBookClickListener {
        void onBookClick(int position, Book book);
    }

    @NonNull
    @Override
    public BookViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.book_layout, parent, false);
        BookViewHolder holder = new BookViewHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull BookViewHolder holder, int position) {
        Book currentBook = bookList.get(position);
        holder.title.setText(currentBook.getTitle());
        holder.bind(listener, currentBook);
    }

    @Override
    public int getItemCount() {
        return bookList.size();
    }

    class BookViewHolder extends RecyclerView.ViewHolder {

        private TextView title;

        public BookViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.tv_title);
        }

        public void bind(final OnBookClickListener listener, final Book book) {
            itemView.setOnClickListener((view) -> listener.onBookClick(getAdapterPosition(), book));
        }
    }
}

File: app/src/main/java/com/gigaworks/tech/bible/ui/bible/adapter/BookAdapter.kt
package com.gigaworks.tech.bible.ui.bible.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.gigaworks.tech.bible.databinding.CardItemBinding
import com.gigaworks.tech.bible.domain.Book

class BookAdapter(
    private val bookList: List<Book>,
    private val clickListener: OnBookClickListener
) : RecyclerView.Adapter<BookAdapter.BookViewHolder>() {

    interface OnBookClickListener {
        fun onBookClick(book: Book)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding = CardItemBinding.inflate(layoutInflater, parent, false)
        return BookViewHolder(binding)
    }

    override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
        holder.bind(bookList[position])
    }

    override fun getItemCount() = bookList.size


    inner class BookViewHolder(private val binding: CardItemBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(book: Book) {
            binding.title.text = book.title
            binding.root.setOnClickListener {
                clickListener.onBookClick(book)
            }
        }

    }
}